File: 3305-count-of-substrings-containing-every-vowel-and-k-consonants-i/3305-count-of-substrings-containing-every-vowel-and-k-consonants-i.java
class Solution {
    public int countOfSubstrings(String word, int k) {
        int dp[] = new int[word.length() + 1], result = 0;
        for (int i = word.length() - 1; i >= 0; i--) {
            if (i == word.length() - 1 || "aeiou".indexOf(word.charAt(i + 1)) < 0) {
                dp[i] = "aeiou".indexOf(word.charAt(i)) < 0 ? 0 : 1;
            } else {
                dp[i] = dp[i + 1] + 1;
            }
        }
        HashMap<Character, Integer> map = new HashMap<>();
        for (int i = 0, j = 0; i < word.length(); i++) {
            for (; j < word.length() && (map.size() < 5 || k > 0); j++) {
                if ("aeiou".indexOf(word.charAt(j)) < 0) {
                    k--;
                } else {
                    map.put(word.charAt(j), map.getOrDefault(word.charAt(j), 0) + 1);
                }
            }
            if (map.size() == 5 && k == 0) {
                result += Math.max(1, dp[j - 1]);
            }
            if ("aeiou".indexOf(word.charAt(i)) < 0) {
                k++;
            } else {
                map.put(word.charAt(i), map.get(word.charAt(i)) - 1);
                if (map.get(word.charAt(i)) == 0) {
                    map.remove(word.charAt(i));
                }
            }
        }
        return result;
    }
}

File: 3305-count-of-substrings-containing-every-vowel-and-k-consonants-i/3305-count-of-substrings-containing-every-vowel-and-k-consonants-i.cpp
class Solution {
public:
    int countOfSubstrings(string word, int K) {
        int n = word.size();

        auto check = [&](char c) {
            return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
        };

        auto calc = [&](int K) {
            long long ret = 0;
            int cnt[26] = {0}, good = 0, x = 0;
            for (int i = 0, j = 0; i < n; i++) {
                while (j < n && (good < 5 || x < K)) {
                    if (check(word[j])) {
                        int t = cnt[word[j] - 'a']++;
                        if (t == 0) good++;
                    } else {
                        x++;
                    }
                    j++;
                }
                if (good == 5 && x >= K) ret += n - j + 1;
                if (check(word[i])) {
                    int t = --cnt[word[i] - 'a'];
                    if (t == 0) good--;
                } else {
                    x--;
                }
            }
            return ret;
        };
        return calc(K) - calc(K + 1);
    }
};